helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/algorithm/ConstraintBasedAlgorithm.java

```diff
@@ -0,0 +1,104 @@
+import org.apache.helix.controller.rebalancer.waged.constraints.SoftConstraint;
+import org.apache.helix.controller.rebalancer.waged.model.AssignableNode;
+import org.apache.helix.controller.rebalancer.waged.model.AssignableReplica;
+import org.apache.helix.controller.rebalancer.waged.model.ClusterContext;
+import org.apache.helix.controller.rebalancer.waged.model.ClusterModel;
+import org.apache.helix.controller.rebalancer.waged.model.OptimalAssignment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+
+/**
+ * The algorithm is based on a given set of constraints
+ * - HardConstraint: Approve or deny the assignment given its condition, any assignment cannot bypass any "hard constraint"
+ * - SoftConstraint: Evaluate the assignment by points/rewards/scores, a higher point means a better assignment
+ * <p>
+ * The goal is to avoid all "hard constraints" while accumulating the most points(rewards) from "soft constraints"
+ */
+public class ConstraintBasedAlgorithm implements RebalanceAlgorithm {
+  private static final Logger LOG = LoggerFactory.getLogger(ConstraintBasedAlgorithm.class);
+  private final List<HardConstraint> _hardConstraints;
+  private final List<SoftConstraint> _softConstraints;
+
+  public ConstraintBasedAlgorithm(List<HardConstraint> hardConstraints, List<SoftConstraint> softConstraints) {
+    _hardConstraints = hardConstraints;
+    _softConstraints = softConstraints;
+  }
+
+  @Override
+  public OptimalAssignment calculate(ClusterModel clusterModel) {
+    OptimalAssignment currentAssignment = new OptimalAssignment();
+    Map<String, Set<AssignableReplica>> replicasByResource = clusterModel.getAssignableReplicaMap();
+    List<AssignableNode> nodes = new ArrayList<>(clusterModel.getAssignableNodes().values());
+    ClusterContext clusterContext = clusterModel.getContext();
+
+    for (Map.Entry<String, Set<AssignableReplica>> resourceToReplicasEntry : replicasByResource.entrySet()) {
+      String resource = resourceToReplicasEntry.getKey();
+      for (AssignableReplica replica : resourceToReplicasEntry.getValue()) {
+        Optional<AssignableNode> bestNode = getNodeWithHighestPoints(replica, nodes, clusterContext, currentAssignment);
+        bestNode.ifPresent(node -> {
+          currentAssignment.addAssignment(replica, node);
+          // adjust the node's capacity and fault zone information
+          node.assign(replica);
+          clusterContext.addPartitionToFaultZone(node.getFaultZone(), resource, replica.getPartitionName());
+        });
+        // stop immediately if any replica cannot find best assignable node
+        if (currentAssignment.hasAnyFailure()) {
```

Comment:
From the code logic, this block could be checked right after line 51. If bestNode is Optional.empty(), currentAssignment has failure. If that happens, we don't need to run the code at line 52.